src/internal/interpreter/http/security/digest/digest.ts

```diff
@@ -0,0 +1,296 @@
+
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   *
+   * @param url url of the request
+   * @param method HTTP method
+   * @param digest extracted of cached digest values
+   * @returns string containing information needed to digest authorization
+   */
+  private buildDigestAuth(
+    url: string,
+    method: string,
+    digest: DigestAuthValues
+  ): string {
+    debugSensitive(
+      `Preparing digest authentication for: ${url} and method: ${method}`
+    );
+    const uri = new URL(url).pathname;
+
+    //Default H1
+    let ha1 = computeHash(
+      digest.algorithm,
+      `${this.configuration.username}:${digest.realm}:${this.configuration.password}`
+    );
+
+    //sess H1 contains original H1 and also nonce and cnonce
+    if (digest.algorithm.endsWith('-sess')) {
+      ha1 = computeHash(
+        digest.algorithm,
+        `${ha1}:${digest.nonce}:${digest.cnonce}`
+      );
+    }
+
+    //H2 is same for default and sess
+    const ha2 = computeHash(digest.algorithm, `${method}:${uri}`);
+    this.nc++;
+    const ncString = String(this.nc).padStart(8, '0');
+
+    let response: string;
+    //Use  QOP
+    if (digest.qop) {
+      //https://datatracker.ietf.org/doc/html/rfc7616#section-3.4.1
+      response = `${ha1}:${digest.nonce}:${ncString}:${digest.cnonce}:${digest.qop}:${ha2}`;
+    } else {
+      response = `${ha1}:${digest.nonce}:${ha2}`;
+    }
+
+    //Hash response
+    const hashedResponse = computeHash(digest.algorithm, response);
+
+    //Build final auth header
+    const opaqueString = digest.opaque ? `opaque="${digest.opaque}"` : '';
+    const qopString = digest.qop ? `qop="${digest.qop}"` : '';
+
+    return [
+      `${digest.scheme} username="${this.configuration.username}"`,
+      `realm="${digest.realm}"`,
+      `nonce="${digest.nonce}"`,
+      `uri="${uri}"`,
+      opaqueString,
+      qopString,
+      `algorithm="${digest.algorithm}"`,
+      `response="${hashedResponse}"`,
+      `nc=${ncString}`,
+      `cnonce="${digest.cnonce}"`,
+    ].join(',');
+  }
+
+  private extractDigestValues(header: string): DigestAuthValues {
+    debugSensitive(`Extracting digest authentication values from: ${header}`);
+
+    const scheme = header.split(/\s/)[0];
+    if (!scheme) {
+      throw missingPartOfDigestHeader(this.challangeHeader, header, 'scheme');
+    }
+    const nonce = extract(header, 'nonce');
+    if (!nonce) {
+      throw missingPartOfDigestHeader(this.challangeHeader, header, 'nonce');
+    }
+
+    return {
+      scheme,
+      algorithm: extractAlgorithm(header),
+      realm: (extract(header, 'realm', false) || '').replace(/["]/g, ''),
+      opaque: extract(header, 'opaque'),
+      qop: extractQop(header),
+      nonce,
+      cnonce: this.makeNonce(),
+    };
+  }
+
+  private makeNonce(): string {
+    let uid = '';
+    for (let i = 0; i < this.cnonceSize; ++i) {
+      uid += this.nonceRaw[Math.floor(Math.random() * this.nonceRaw.length)];
```

Comment:
I'm not sure about the `Math.random` security implications here.

@lukas-valenta What do you think?

Reply:
On the other hand, it is only used to compute cnonce.. 🤷

Reply:
i missed that. cnonce or no, let's use crypto